### DISCUSSION TITLE
Buffer with preserve topology option

### DISCUSSION THREAD

**Discussion by StefanKarlsson987** [shapely/shapely#1739] [Q&A]
I wonder if it would be possible to have an option for shapely.buffer to not apply the final union operation.  With such option enabled, invalid geometries would result, i.e. with intersections. 

The reason we would want to have such option is for when buffer is used in mathematical morphology. One could want to have an invalid geometry as part of a step to arrive at valid geometry. 

code example below for an example. More details outlined [here](https://stackoverflow.com/questions/75272295/shapely-smoothing-with-dilate-and-erode) 

```
from shapely.geometry import Polygon
import numpy as np
from descartes import PolygonPatch
from matplotlib import pyplot as plt

# create a large square
x = np.array([-5,-5,5,5])
y = np.array([5,-5,-5,5])
poly1 = Polygon(zip(x,y))

# smaller square to cut away from first
x = np.array([-4,-4,4,4])
y = np.array([4,-4,-4,4])
poly2 = Polygon(zip(x,y))

# small shape to cut away from left side
x = np.array([-11,-11,0,0])
y = np.array([1,-1,0,0])
poly3 = Polygon(zip(x,y))

poly_t=poly1.difference(poly2)
poly4 = poly_t.difference(poly3)

poly5= poly4.buffer(0.45)
poly5= poly5.buffer(-0.45)

fig = plt.figure()

ax = fig.add_subplot(121)
plt.axis([-5.5, 5.5, -5.5, 5.5])

patch = PolygonPatch(poly5)
ax.add_patch(patch)

plt.show()
```

**Comment by StefanKarlsson987** [comment]
For sake of library architecture, it could be a valid option to implement nesting of these operations within a library construct.

If one could define something like 
nestedoperation(("buffer",val1, otherArgs..),("buffer",val2, otherArgs..)
then shapely could still ensure valid geometry output at the end. 

Just thinking aloud. Personally, I dont see the harm of having self-intersections if the user specifically asks for it.

**Comment by jorisvandenbossche** [comment]
The actual Buffer operation is performed by the GEOS library (https://github.com/libgeos/geos), so this is probably a question to ask there. 

I can't directly think of an easy workaround with the current functionality in Shapely / GEOS (except for manually cutting your polygon in two pieces, and do the buffer for each of the parts (avoiding the union of the result that overlaps), but that would of course be very custom to this specific polygon and difficult to generalize). The `offset_curve` function has the same behaviour as `buffer`.

**Reply by StefanKarlsson987** [reply]
Thanks for reply and assistance.

I will try to ask for this in the GEOS repo, as you suggested. Could I ask your opinion of whether this is a reasonable feature to request?
Do you agree that this would be a desirable feature to incorporate into Shapely? In particular, if GEOS implemented a "preserve topology" option in their c++ buffer function, Shapely would be quick to make that interface available as well.

**Reply by StefanKarlsson987** [reply]
issue written for GEOS, find it [here](https://github.com/libgeos/geos/issues/813)
